### PR TITLE
Add board MikroElektronika Quail

### PR DESCRIPTION
**Pull request by lorcap**
[Quail](https://www.mikroe.com/quail) (PID: MIKROE-1793) is based on STM32F427VI CPU, featuring 2048 kB of Flash memory and 192 kB of RAM. An on-board Cypress S25FL164K adds 8 MB of SPI Flash.

Quail has 4 mikroBUS™ sockets for Mikroe click™ board connectivity, along with 24 screw terminals for connecting additional electronics and two USB ports (one for programming, the other for external mass storage).

4 UARTs, 2 SPIs and 1 I²C bus are available for communication.

PS: this PR adds more details to Clicker2, another MikroE board. I hope you don't mind if I mixed up things. A PR just for that was a bit overkilling for me.

### CODE REVIEW

File: ports/stm32/boards/MIKROE_QUAIL/board.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    "docs": "",
+    "features": [
+        "mikroBUS"
+    ],
+    "id": "MIKROE-QUAIL",
+    "images": [
+            "quail_hand.jpg",
+            "quail_top.jpg",
+            "quail_bot.jpg",
+            "quail_bus.jpg"
```
Comment (dpgeorge) [inline ports/stm32/boards/MIKROE_QUAIL/board.json:14]: please submit these images to https://github.com/micropython/micropython-media

also, probably only need one image and `quail_top.jpg` looks the most suitable

File: ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h [hunk 2]
```diff
@@ -0,0 +1,113 @@
+#define MICROPY_HW_BOARD_NAME       "MikroE Quail"
+#define MICROPY_HW_MCU_NAME         "STM32F427VI"
+#define MICROPY_PY_SYS_PLATFORM     "MikroE Quail"
```
Comment (dpgeorge) [inline ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h:3]: This line should probably be removed so the setting stays as the default of `pyboard`.

File: ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+#define MICROPY_HW_CLK_LAST_FREQ    (1)
+
+// The board has no crystal for the RTC
+#define MICROPY_HW_RTC_USE_LSE      (0)
+#define MICROPY_HW_RTC_USE_US       (0)
+#define MICROPY_HW_RTC_USE_CALOUT   (0)  // turn on/off PC13 512Hz output
+
+// UART config
+// mikroBUS slot 1
+#define MICROPY_HW_UART3_NAME       "SLOT1"
```
Comment (dpgeorge) [inline ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h:28]: Are these good names, do they match something written on the board itself?

File: ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.mk [hunk 4]
```diff
@@ -0,0 +1,13 @@
+MCU_SERIES = f4
+CMSIS_MCU = STM32F427xx
+AF_FILE = boards/stm32f427_af.csv
+ifeq ($(USE_MBOOT),1)
```
Comment (dpgeorge) [inline ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.mk:4]: Did you test with mboot?

File: ports/stm32/boards/MIKROE_QUAIL/pins-zerynth.csv [hunk 5]
```diff
@@ -0,0 +1,211 @@
+# Pin mapping according to Zerynth documentation
```
Comment (dpgeorge) [inline ports/stm32/boards/MIKROE_QUAIL/pins-zerynth.csv:1]: I'm not sure this file should be included, there's really no reason to match Zerynth, code from there will not run in MicroPython.

Pins should be named based on the labels on the physical board and/or the board documentation.

File: ports/stm32/boards/stm32f427_af.csv [hunk 6]
```diff
@@ -0,0 +1 @@
+stm32f429_af.csv
```
Comment (dpgeorge) [inline ports/stm32/boards/stm32f427_af.csv:1]: It looks like this is a symlink that was added as a file.  Symlinks should not be used.  Best to just remove this file, and reference `stm32f429_af.csv` in the `mpconfigboard.mk` file directly.

**Review by dpgeorge — CHANGES_REQUESTED** [review]

File: ports/stm32/boards/MIKROE_QUAIL/board.json (hunk 1 above)
Comment (lorcap) [reply ports/stm32/boards/MIKROE_QUAIL/board.json]: OK, done.

File: ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h (hunk 2 above)
Comment (lorcap) [reply ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h:3]: Just the last one or all of them? It'd be nice to see the first two board references when running `mpremote`.

File: ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h (hunk 3 above)
Comment (lorcap) [reply ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h:28]: Well, as you can see from the [picture](https://github.com/micropython/micropython-media/blob/143855a2f0852f87fe9b5037d6a61488747888e3/boards/MIKROE_QUAIL/quail_top.jpg), each slot is named "mikroBUS" plus a big number 1 to 4. `SLOT1` is concise and pretty obvious, I believe. I was inspired by another MikroE board supported by MicroPython, [CLICKER2](https://github.com/micropython/micropython/tree/master/ports/stm32/boards/MIKROE_CLICKER2_STM32).

I don't mind change the name. What would you suggest? `MIKROBUS1`, `MIKRO1`, `BUS1`, `MB1`?

File: ports/stm32/boards/MIKROE_QUAIL/pins-zerynth.csv (hunk 5 above)
Comment (lorcap) [reply ports/stm32/boards/MIKROE_QUAIL/pins-zerynth.csv:1]: OK, I'll do that.

File: ports/stm32/boards/stm32f427_af.csv (hunk 6 above)
Comment (lorcap) [reply ports/stm32/boards/stm32f427_af.csv:1]: I'll do it, even though I find it a bit misleading. [Datasheet](https://www.st.com/resource/en/datasheet/stm32f427vi.pdf) says (table 12 on page 75) that the alternate functions are exactly the same for the two microcontrollers. Shall we rename the old `stm32f429_af.csv` to something like `stm32f4279_af.csv`? At the moment there are 3 more boards relaying on the same file: `NUCLEO_F446RE`, `NUCLEO_F429ZI`, and `STM32F429DISC`.

File: ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.mk (hunk 4 above)
Comment (lorcap) [reply ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.mk:4]: No, I didn't. I've just tried to compile it and it fails. I'll disable it for now. I'll come back to it because I need it in my project.

File: ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h (hunk 2 above)
Comment (dpgeorge) [reply ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h:3]: Just the last one.  You'll still see the banner with the full board name and MCU.

File: ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h (hunk 3 above)
Comment (dpgeorge) [reply ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h:28]: Ah, the same names are used by the CLICKER2 board definition.  So, yes, keep them as they are.

Comment (lorcap) [reply ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h:28]: OK.

File: ports/stm32/boards/stm32f427_af.csv (hunk 6 above)
Comment (dpgeorge) [reply ports/stm32/boards/stm32f427_af.csv]: There are quite a few af.csv files that are used by multiple MCU series.  Let's leave the file name as-is and I suggest adding a comment in the mpconfigboard.mk file to mention that the f429 file is used because it's equivalent.

File: ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h (hunk 2 above)
Comment (lorcap) [reply ports/stm32/boards/MIKROE_QUAIL/mpconfigboard.h:3]: OK.

File: ports/stm32/boards/MIKROE_QUAIL/board_init.c [hunk 7]
```diff
@@ -0,0 +1,11 @@
+#include "py/mphal.h"
+
+void board_early_init(void) {
+
+#if !MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE
+    // set external SPI flash CS pin high
+    mp_hal_pin_output(MICROPY_HW_SPIFLASH_CS);
+    mp_hal_pin_write(MICROPY_HW_SPIFLASH_CS, 1);
```
Comment (dpgeorge) [inline ports/stm32/boards/MIKROE_QUAIL/board_init.c:8]: Why is this needed?  The CS pin should be configured as output and set high by default in the SPI flash ioctl init code.

File: ports/stm32/boards/stm32f427_af.csv (hunk 6 above)
Comment (lorcap) [reply ports/stm32/boards/stm32f427_af.csv]: OK. Done.

File: ports/stm32/boards/MIKROE_QUAIL/board_init.c (hunk 7 above)
Comment (lorcap) [reply ports/stm32/boards/MIKROE_QUAIL/board_init.c]: I just put together what I found in other ports without diving too much into the inner gears. This one comes from Sparkfun Micromod's [board_init.c](https://github.com/micropython/micropython/blob/aecb697c72ac1fdb9f549daa5734e2ffbd37a6d3/ports/stm32/boards/SPARKFUN_MICROMOD_STM32/board_init.c).

When removed, I can still see the filesystem on the external FLASH. Hence, yes, it's not required.

**Comment by dpgeorge** [comment]
Thanks for updating.  Merged in 16c7a808742570e3c9ab1e1f8243d0a2166995a3 and 8f0e304e65a83e1f52dfb29185159e8eb9f7f34d
